### PR TITLE
New schema scripts

### PR DESCRIPTION
**Pull request by cavis**
🚧 WIP

For PRX/dovetail.prx.org#264

Schema changes:

- `dt_downloads`
  - add `remote_referrer` (should have been here already)
  - add `latitude`, `longitude`, `postal_code`
  - rename `city_id` to `city_geoname_id` and `country_id` to `country_geoname_id`
  - add `clienthash`, the sha256 of path+rawip+agent that DT computes (recomputed during this migration)
  - `remote_ip` is now masked to omit the last octet(s)
- `dt_impressions`
  - drop many things redundant with downloads table (digest, program, path, remote_agent, remote_ip, city_id, country_id, agent_name_id, agent_type_id, agent_os_id)
  - keeps 4 redundant fields for easy access (feeder_podcast, feeder_episode, is_duplicate, cause)

### CODE REVIEW

File: scripts/create_tables.rb [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  table.description = 'Dovetail downloads'
+  table.time_partitioning_type = 'DAY'
+  table.time_partitioning_field = 'timestamp'
+  table.schema do |schema|
+    schema.timestamp 'timestamp', mode: :required
+    schema.string 'request_uuid', mode: :required
+
+    # redirect data
+    schema.integer 'feeder_podcast'
+    schema.string 'feeder_episode'
```
[the remark this answers is not in the document]
Comment (cavis) [reply scripts/create_tables.rb]: yeah, it's changing/removing columns that's a pain.  adding... no problem!

File: scripts/create_tables.rb [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    schema.string 'program'
+    schema.string 'path'
+    schema.string 'clienthash'
+    schema.string 'digest'
```
[the remark this answers is not in the document]
Comment (cavis) [reply scripts/create_tables.rb]: yep, digest is the unique key for that arrangement of files.

[the remark this answers is not in the document]
Comment (cavis) [reply scripts/create_tables.rb]: and clienthash is new ... it's what DT uses to do campaign/recent-redirect de-duping.  going to start sending that to BQ, to represent our concept of a "unique user" (path+ip+agent).

File: scripts/create_tables.rb [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+    # derived data
+    schema.integer 'agent_name_id'
+    schema.integer 'agent_type_id'
+    schema.integer 'agent_os_id'
+    schema.integer 'city_geoname_id'
+    schema.integer 'country_geoname_id'
+    schema.string 'postal_code'
+    schema.float 'latitude'
+    schema.float 'longitude'
```
[the remark this answers is not in the document]
Comment (cavis) [reply scripts/create_tables.rb]: DMA is actually in the `geonames` table.  So what I've been doing thus far is (1) group downloads by guid/time/city_id, (2) join the geonames table on city_id, (3) group by geonames.metro_code.  It's been decently fast.  If it's ever not fast enough ... I guess we'll have to denormalize some of geonames further onto this table?

File: scripts/create_tables.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (kookster) [reply scripts/create_tables.rb]: yeah the new thing I saw in the PR desc, and we had discussed, just wanted to make sure I knew what digest was - thanks!

File: scripts/create_tables.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (kookster) [reply scripts/create_tables.rb]: I see, that makes sense - if it performs for now, I got no issue. 👍

**Pull request merged**